[DOC] Tracing: Add doc for Tempo Search query (#68923)
grafana/grafana: docs/sources/shared/datasources/tempo-search-traceql.md

---
headless: true
---

> <!--
> This file documents the Search query type for the Tempo data source. It is available as a public preview.
> 
> This shared file is included in these locations:
> /grafana/docs/sources/datasources/tempo/query-editor/index.md
> 
> If you make changes to this file, verify that the meaning and content are not changed in any place where the file is included.
> 
> Any links should be fully qualified and not relative: /docs/grafana/
>  -->

Author: ```suggestion

[//]: # (This file documents the Search query type for the Tempo data source. It is available as a public preview.)

[//]: # (This shared file is included in these locations:)
[//]: # (/grafana/docs/sources/datasources/tempo/query-editor/index.md)

[//]: # (If you make changes to this file, verify that the meaning and content are not changed in any place where the file is included.)

[//]: # (Any links should be fully qualified and not relative: /docs/grafana/) 
```

---
headless: true
---

<!--
This file documents the Search query type for the Tempo data source. It is available as a public preview.

This shared file is included in these locations:
/grafana/docs/sources/datasources/tempo/query-editor/index.md

If you make changes to this file, verify that the meaning and content are not changed in any place where the file is included.

Any links should be fully qualified and not relative: /docs/grafana/

>  -->

Reviewer: Nice! Really helpful comment.

Another place where we can consider using "Markdown comments" instead of HTML comments to avoid this ending up the page HTML.

Author: Thank you. I've updated the comments to use the Markdown comments you highlighted earlier.

# Create TraceQL queries using Search

> {{% admonition type="Note" %}}

Reviewer: ```suggestion
{{% admonition type="note" %}}
```

Admonition types should be specified in lower case.

Author: Thank you. I had some trouble with the admonition rendering and was playing with it to see if it made a difference. I built it today and things seem to be okay now.

> This new Search query type is available as an experimental feature. This feature is under active development and is not production ready. To try this feature, enable the `traceqlSearch` feature flag in Grafana (read documentation). Grafana Cloud users should contact Grafana Support to enable this feature.

Reviewer: I'm not sure about the parenthetical here, it doesn't seem that helpful to a reader.

Author: Because the link is missing :)

Author: ```suggestion
This new Search query type is available as an experimental feature. This feature is under active development and is not production ready. To try this feature, enable the `traceqlSearch` feature flag in Grafana (read [documentation](/docs/grafana/latest/setup-grafana/configure-grafana/feature-toggles/)). Grafana Cloud users should contact Grafana Support to enable this feature.
```

You can use the query builder to search trace data by resource service name, span name, duration, and one or more tags. The examples on this page use the default filters.

In addition, you can add query builder blocks, view the query history, and use the **Inspector** to see details.

{{< figure src="/static/img/docs/queries/screenshot-tempods-query-search.png" class="docs-image--no-shadow" max-width="750px" caption="Screenshot of the Tempo Search query type" >}}

## Perform a search

To perform a search, you need to select filters and/or tags and then run the query. The results appear underneath the query builder.
The screenshot below identifies the areas used to perform a search.

{{< figure src="/static/img/docs/queries/screenshot-tempods-query-search-parts.png" class="docs-image--no-shadow" max-width="750px" caption="Parts of Tempo Search query type" >}}

|Number|Name|Action|Comment|
|:----|:----|:----|:----|
|1|Data source|Use the data source drop-down to select a Tempo data source.|Each data source has its own version of search. This Search is specific to the Tempo data source.|
|2|Query type tab|Select Search.| |

> |3|Choose filter|Choose whether to filter using **Resource Service Name**, **Span Name**, and/or **Duration**.|You must have at least one filter or tag defined to run the query.|

Author: ```suggestion
|3|Choose filter|Choose whether to filter using **Resource Service Name**, **Span Name**, and/or **Duration**.|You can execute an empty query in the Search tab. In TraceQL, `{}` is a valid query and is the default query until you choose options.|
```

> |3|Choose filter|Choose whether to filter using **Resource Service Name**, **Span Name**, and/or **Duration**.|You must have at least one filter or tag defined to run the query.|

Author: Thank you. I've updated the text to reflect this.

Using filters, you refine the data returned from the query by selecting Resource Service Name, Span Name, or Duration. The **Duration** represents span time, calculated by subtracting the end time from the start time of the span.

Grafana administrators can change the default filters using the Tempo data source configuration.
Filters can be limited by the operators. The available operators are determined by the field type.
For example, **Span Name** and **Resource Service Name** are string fields so the comparison operators are equals (`=`), not equal (`!=`), or regular expressions (`=~`).
**Duration** is a duration field type and uses range selections (`>`, `>=`, `<`, `<=`).

You must define at least one filter or one tag for a query to run.

To define filters, follow these steps:

1. Choose one of the filters.
1. Select a comparison operator from the drop-down.
1. **Resource Service Name** and **Span Name** only: Select one or more values from the drop-down.
1. **Duration** only: Enter values and units for the range and choose comparison operators for the drop-downs. Units can be nanoseconds (`ns`), milliseconds (`ms`), seconds (`s`), minutes (`m`), and hours (`h`).

> 

Author: Thank you for this information. I've added the content to line 60 above, so it's in the paragraph with duration.

> 

Author: Thank you for the great info! I've added another paragraph. What do you think?

Every query searches the data for the selected time frame.
By default, queries run against data from the last hour.
Select Time range to the left of Run query to choose the time range for the data your query runs against.
Read the [dashboard time range](/docs/grafana/latest/dashboards/use-dashboards/#set-dashboard-time-range) documentation to learn more.

To access Search, use the following steps:

1. Sign into Grafana.
1. Select your Tempo data source.
1. From the menu, choose Explore and select Query type > Search.

## Define filters

Using filters, you refine the data returned from the query by selecting Resource Service Name, Span Name, or Duration. The **Duration** represents span time, calculated by subtracting the end time from the start time of the span.

Grafana administrators can change the default filters using the Tempo data source configuration.
Filters can be limited by the operators. The available operators are determined by the field type.
For example, **Span Name** and **Resource Service Name** are string fields so the comparison operators are equals (`=`), not equal (`!=`), or regular expressions (`=~`).

> **Duration** is a duration field type and uses range selections (`>`, `>=`, `<`, `<=`).

Author: ```suggestion
**Duration** is a duration field type and uses range selections (`>`, `>=`, `<`, `<=`). When you select multiple values for the same tag, Grafana automatically changes the operator to the regex operator `=~` and concatenates the values with a `|`.
```

Every query searches the data for the selected time frame.
By default, queries run against data from the last hour.
Select Time range to the left of Run query to choose the time range for the data your query runs against.
Read the [dashboard time range](/docs/grafana/latest/dashboards/use-dashboards/#set-dashboard-time-range) documentation to learn more.

To access Search, use the following steps:

1. Sign into Grafana.
1. Select your Tempo data source.
1. From the menu, choose Explore and select Query type > Search.

## Define filters

Using filters, you refine the data returned from the query by selecting Resource Service Name, Span Name, or Duration. The **Duration** represents span time, calculated by subtracting the end time from the start time of the span.

Grafana administrators can change the default filters using the Tempo data source configuration.
Filters can be limited by the operators. The available operators are determined by the field type.
For example, **Span Name** and **Resource Service Name** are string fields so the comparison operators are equals (`=`), not equal (`!=`), or regular expressions (`=~`).

> 
> **Duration** is a duration field type and uses range selections (`>`, `>=`, `<`, `<=`). When you select multiple values for the same tag, Grafana automatically changes the operator to the regex operator `=~` and concatenates the values with a `|`.

Author: ```suggestion
When you select multiple values for the same tag, Grafana automatically changes the operator to the regex operator `=~` and concatenates the values with a `|`.

**Duration** is a duration field type and uses range selections (`>`, `>=`, `<`, `<=`). 
```

Author: Line 67 should be in a code block. The formatting in the suggestion box doesn't like the additional code block. I'll fix it once we get the suggestion wording right.

Author: @adrapereira How does this update look?
